BC1.2 charging mode support

keith-zephyr opened the pull request:
Add support for BC1.2 peripherals to act as the charging device. This is a capability often provided on USB Type-C ports.

The BC1.2 charging mode advertised is configured with a devicetree property.

keith-zephyr committed:
test: bc12: rename test

Rename the main test file to pd_mode.c as the test verfies portable
device mode only.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
bc12: Add charging mode role information to callback

Update the result callback routine to also provide information about the
charging mode role. This permits tests to validate plug/unplug events
detected by the BC1.2 device operating in charging mode.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
bc12: Add charging mode support

Add chargingg mode support to the BC1.2 API and the PI3USB9201 driver.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
emul: pi3usb9201: Add charging mode support

Add charging mode support the PI3USB9201 emulator and add a backend API
for connecting/disconnecting a portable device partner.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
test: bc12: add charging mode test

Add chargning mode test for the pi3usb9201 BC1.2 device.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
docs: bc12: Update for charging mode operation

Add details for using charging mode and add the BC1.2 emulator API
information to the document.

Signed-off-by: Keith Short <keithshort@google.com>

jfischer-no commented on drivers/usb/bc12/bc12_pi3usb9201.c:
```diff
@@ -148,22 +148,49 @@ static void pi3usb9201_notify_callback(const struct device *dev,
 		pi3usb9201_notify_callback(dev, state);
 	} else {
-		pi3usb9201_data->partner_type = BC12_TYPE_NONE;
+		pi3usb9201_data->partner_state.bc12_role = BC12_DISCONNECTED;
```
I found the use of the state parameter a bit confusing, NULL is passed only in pi3usb9201_disconnect() to cleanup, right?

keith-zephyr replied:
Added a description of the function and parameters.

jfischer-no commented on drivers/usb/bc12/bc12_pi3usb9201.c:
```diff
@@ -110,6 +122,24 @@ static int pi3usb9201_set_mode(const struct device *dev, enum pi3usb9201_mode mo
 				      mode << PI3USB9201_REG_CTRL_1_MODE_SHIFT);
 }
 
+static int pi3usb9201_get_mode(const struct device *dev, enum pi3usb9201_mode *mode)
```
```suggestion
static int pi3usb9201_get_mode(const struct device *dev, enum pi3usb9201_mode *const mode)
```

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/0d0c6d44d3480b8293c0277d36f169b7d4b13af5 is a rebase only.

alevkoy commented on tests/drivers/bc12/src/pd_mode.c:
```diff
@@ -19,15 +19,15 @@
 #include <zephyr/logging/log.h>
 
-#define LOG_MODULE_NAME test_bc12
+#define LOG_MODULE_NAME test_bc12_pd
```
Should this be in the previous commit? Same question for some of the other name changes below.

alevkoy commented on drivers/usb/bc12/bc12_pi3usb9201.c:
```diff
@@ -386,7 +416,7 @@ static int pi3usb9201_init(const struct device *dev)
 	 * Set most recent bc1.2 detection type result to
 	 * BC12_TYPE_NONE for the port.
```
Update comment?

keith-zephyr replied:
Done

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/d12b2ff3d9ed366a6d63188aa94c513421083d3c is a rebase only.

keith-zephyr committed:
test: bc12: cleanup naming

Cleanup some internal names.  Remove references to the pi3usb9201 device
as the test should work with all BC1.2 driver/emulators.

Signed-off-by: Keith Short <keithshort@google.com>

jfischer-no commented on drivers/usb/bc12/bc12_pi3usb9201.c:
```diff
@@ -148,22 +148,49 @@ static void pi3usb9201_notify_callback(const struct device *dev,
 }
 
+static bool pi3usb9201_partner_has_changed(const struct device *dev,
+					   struct bc12_partner_state *state)
```
This is not resolved but is actually very helpful here.
